### PR TITLE
Avoid clearing sequential bits in a loop

### LINKED ISSUE #54: Missing check for message limits, issues with data message counter
The protocol defines the following message limits:
* Rekey-After-Messages 2^64 * 2^16 - 1
* Reject-After-Messages 2^64 − 2^4 − 1

Although unlikely in practice, the implementation must check that these message limits are not reached and avoid a counter wraparound. The implementation of ReceivingKeyCounterValidator in src/noise/session.rs must be updated to check this.

Right now a malicious implementation could produce the following messages sequence which is accepted by boringtun:

* counter=0 (sets next to 1 in `mark_did_receive`)
* counter=2^64 - 1 (sets next to 0 due to integer overflow)
* counter=0 (oops! counter reuse - replay accepted!)

Additionally, going from counter 0 to counter 2^64 will also result in a lot of unnecessary executions which facilitates a DoS:
```rust
        // Packets where dropped, or maybe reordered, skip them and mark unused
        let mut i = self.next; // 1
        while i < counter { // i = 1 .. 2^64 - 1
            self.clear_bit(i);
            i += 1;
        }
```
If the gap is too big, perhaps the whole bitmap should be cleared.

FWIW, the WireGuard kernel implementation uses a window of 2048 bits whereas this one uses 1024.

Another related issue is that `sending_key_counter` and `sending_key_counter` are an [AtomicUsize](https://doc.rust-lang.org/std/sync/atomic/struct.AtomicUsize.html) which may be 32-bit on 32-bit architectures whereas the counter in the protocol is 64-bit.

### PR DESCRIPTION
**Pull request by vkrasnov**
Fixes #54

### CODE REVIEW

File: src/noise/session.rs [hunk 1]
```diff
@@ -81,6 +81,14 @@ impl ReceivingKeyCounterValidator {
         self.bitmap[word] &= !(1u64 << bit);
     }
 
+    // Clear the world that contains idx
```
Comment (Lekensteyn) [inline src/noise/session.rs]: typo: "world" -> "word"

File: src/noise/session.rs [hunk 2]
```diff
@@ -128,10 +136,22 @@ impl ReceivingKeyCounterValidator {
...
+        if counter - self.next >= N_BITS {
+            // Too far ahead, clear all the bits
+            for c in self.bitmap.iter_mut() {
+                *c = 0;
+            }
+        } else {
+            let mut i = self.next;
+            while i + WORD_SIZE < counter {
+                // Clear whole word at a time
+                self.clear_word(i);
```
Comment (Lekensteyn) [inline src/noise/session.rs]: Won't this also clear bits before and including "next" if "next" does not align on a word boundary?

I think that another loop is needed before:
```rust
while i % WORD_SIZE != 0 && i < counter {
    self.clear_bit(i);
    i += 1;
}
```

(+need test case)

Comment (vkrasnov) [reply src/noise/session.rs]: Fixed now

**Comment by Lekensteyn** [comment]
LGTM. Would you consider adding a check for counters past the limits in #54 though? As you said it won't happen in practice, but since it is not permitted by the spec it should not hurt to block it either.

**Comment by vkrasnov** [comment]
I think the implementation (or did I see it somewhere else?) says that the numbers are a placeholder in case a realistic limit is needed. But since those are hardcoded, the value can only change with the implementation anyway

**Comment by Lekensteyn** [comment]
At least the Linux kernel implementation does have an explicit check for those insanely large numbers. In the protocol spec (whitepaper), the counter starts at 0 and increases for every data packet. So it seems reasonable to me to check the counter against the REJECT-AFTER-MESSAGES limit.

**Pull request merged**